src/device-state.ts

```diff
@@ -0,0 +1,890 @@
+		]);
+		this.maxPollTime = conf.appUpdatePollInterval;
+
+		await this.applications.init();
+		if (!conf.initialConfigSaved) {
+			return this.saveInitialConfig();
+		}
+
+		this.initNetworkChecks(conf);
+
+		log.info('Reporting initial state, supervisor version and API info');
+		await this.reportCurrentState({
+			api_port: conf.listenPort,
+			api_secret: conf.apiSecret,
+			os_version: conf.osVersion,
+			os_variant: conf.osVariant,
+			supervisor_version: conf.version,
+			provisioning_progress: null,
+			provisioning_state: '',
+			status: 'Idle',
+			logs_channel: null,
+			update_failed: false,
+			update_pending: false,
+			update_downloaded: false,
+		});
+
+		const targetApps = await this.applications.getTargetApps();
+		if (!conf.provisioned || (_.isEmpty(targetApps) && !conf.targetStateSet)) {
+			try {
+				await loadTargetFromFile(null, this);
+			} finally {
+				await this.config.set({ targetStateSet: true });
+			}
+		} else {
+			log.debug('Skipping preloading');
+			if (conf.provisioned && !_.isEmpty(targetApps)) {
+				// If we're in this case, it's because we've updated from an older supervisor
+				// and we need to mark that the target state has been set so that
+				// the supervisor doesn't try to preload again if in the future target
+				// apps are empty again (which may happen with multi-app).
+				await this.config.set({ targetStateSet: true });
+			}
+		}
+		await this.triggerApplyTarget({ initial: true });
+	}
+
+	public async initNetworkChecks({
+		apiEndpoint,
+		connectivityCheckEnabled,
+	}: {
+		apiEndpoint: ConfigType<'apiEndpoint'>;
+		connectivityCheckEnabled: ConfigType<'connectivityCheckEnabled'>;
+	}) {
+		network.startConnectivityCheck(
+			apiEndpoint,
+			connectivityCheckEnabled,
+			connected => {
+				return (this.connected = connected);
+			},
+		);
+		this.config.on('change', function(changedConfig) {
+			if (changedConfig.connectivityCheckEnabled != null) {
+				return network.enableConnectivityCheck(
+					changedConfig.connectivityCheckEnabled,
+				);
+			}
+		});
+		log.debug('Starting periodic check for IP addresses');
+
+		await network.startIPAddressUpdate()(async addresses => {
+			await this.reportCurrentState({
+				ip_address: addresses.join(' '),
+			});
+		}, constants.ipAddressUpdateInterval);
+	}
+
+	private async saveInitialConfig() {
+		const devConf = await this.deviceConfig.getCurrent();
+
+		await this.deviceConfig.setTarget(devConf);
+		await this.config.set({ initialConfigSaved: true });
+	}
+
+	// We keep compatibility with the StrictEventEmitter types
+	// from the outside, but within this function all hells
+	// breaks loose due to the liberal any casting
+	private emitAsync<T extends keyof DeviceStateEvents>(
+		ev: T,
+		...args: DeviceStateEvents[T] extends (...args: any) => void
+			? Parameters<DeviceStateEvents[T]>
+			: Array<DeviceStateEvents[T]>
+	) {
+		if (_.isArray(args)) {
+			return setImmediate(() => this.emit(ev as any, ...args));
```

Comment:
Why the `setImmediate`s?

Reply:
I'm not sure, this was part of the code from long before I took over the supervisor.